vcd-cli 210: Fixed KeyError and added system test for search

hodgesrm opened the pull request:
Fixed KeyError that appears if you search for an entity type that has
no corresponding entities. Added docstring to Client.get_typed_query
method to make it clearer what arguments do. Implemented system test for
search to verify fix as well common search API usage.  Updated README.md
in system_tests directory to explain how to develop new tests.

(This arose in vcd-cli testing, hence the reference to a vcd-cli issue.)

Reviewers: @ckolovson, @rocknes

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vmware/pyvcloud/245)
<!-- Reviewable:end -->

rocknes commented:
Travis build has failed for this PR.

sahithi commented:
Reviewed 1 of 3 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[system_tests/README.md, line 40 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDDSJtb5RrmV0IOI6oc:-LDDSJtb5RrmV0IOI6od:b-1bmtrn) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/README.md#L40)):*
> ```Markdown
> Step 1: Start by copying an existing test to a new name such as my_test.py.
> 
> Step 2: Trim out extraneous code and add your own test cases.  Name the
> ```

Can we mention about the naming convention as well for public test methods and private helper methods?

---

*[system_tests/README.md, line 67 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDDSUl3EreYW_iAhovT:-LDDSUl3EreYW_iAhovU:bhtqg1y) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/README.md#L67)):*
> ```Markdown
> something useful is missing. 
> 
> Step 5: Fill out values in the base_config.yaml file to tell your
> ```

Do we have sample base_config.yaml for reference?

---


*Comments from [Reviewable](https://reviewable.io/reviews/vmware/pyvcloud/245)*
<!-- Sent from Reviewable.io -->

ckolovson commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[system_tests/README.md, line 51 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDDZsO4EwlVDE8NL7dB:-LDDZsO4EwlVDE8NL7dC:b-dhfnql) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/README.md#L51)):*
> ```Markdown
> 
> Warning: If you unittest setup and teardown methods be aware that
> these run on every case.  If they do anything complex your test will be
> ```

Not sure what you’re trying to say here. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/vmware/pyvcloud/245)*
<!-- Sent from Reviewable.io -->

rocknes commented:
Reviewed 1 of 3 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[system_tests/README.md, line 51 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDDZsO4EwlVDE8NL7dB:-LDE_bxg6xqArZ9Y4uKE:bh6qiqm) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/README.md#L51)):*
<details><summary><i>Previously, ckolovson (Curt Kolovson) wrote…</i></summary><blockquote>

Not sure what you’re trying to say here. 
</blockquote></details>

I guess he meant," if you _use_ unittest setup and teardown methods..."

---

*[system_tests/README.md, line 67 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDDSUl3EreYW_iAhovT:-LDDyE7m5LYp2_mjYR7X:b4ju393) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/README.md#L67)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Do we have sample base_config.yaml for reference?
</blockquote></details>

Yes sample base_config.yaml has been checked in git repo.

---

*[system_tests/search_tests.py, line 23 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDEaLfS3E4sSLtO_GeR:-LDEaLfS3E4sSLtO_GeS:bfshn7o) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/search_tests.py#L23)):*
> ```Python
> 
> from pyvcloud.vcd.client import QueryResultFormat
> from pyvcloud.vcd.client import ResourceType
> ```

I think we should get rid of RESOURCE_TYPES and not use it. 
It's not being used anywhere else in the codebase. While introducing ResourceType enum, Curt kept RESOURCE_TYPE list just to be on the safe side.

---

*[system_tests/search_tests.py, line 31 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDEaurf9QPcTkIlhXCR:-LDEaurf9QPcTkIlhXCS:b-dw43dr) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/search_tests.py#L31)):*
> ```Python
>     """Test pyvcloud search functions"""
> 
>     def setUp(self):
> ```

This will run before every test. Is that the intent here?

---

*[system_tests/search_tests.py, line 48 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDEbS0nCTblirsv3wfI:-LDEbS0nCTblirsv3wfJ:bbs8hu4) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/search_tests.py#L48)):*
> ```Python
>             len(q1_records) > 0,
>             msg="Find at least one organization")
>         for r in q1_records:
> ```

Why not just use
name0 = q1_records[0] .name?

---

*[system_tests/search_tests.py, line 72 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDEcH1-5p9XEvQpFb7s:-LDEcH1-5p9XEvQpFb7t:b-hjqp33) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/search_tests.py#L72)):*
> ```Python
>         self.assertEqual(1, len(q3_records), msg="Find org with query filter")
> 
>     def test_0020_find_existing_with_org_user(self):
> ```

.._find_existing_entities_..  ?

---

*[system_tests/search_tests.py, line 110 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDEdQ4q45OK-TBpJXMi:-LDEdQ4q45OK-TBpJXMj:b-qvficr) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/search_tests.py#L110)):*
> ```Python
>                 resource_type,
>                 query_result_format=QueryResultFormat.ID_RECORDS)
>             q1_records = list(q1.execute())
> ```

Have you run this test locally?
We don't have resources like TO_CLOUD_TUNNEL in the testbed.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vmware/pyvcloud/245)*
<!-- Sent from Reviewable.io -->

sahithi commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> after addressing other outstanding comments

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[system_tests/search_tests.py, line 65 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDIAJnU1NJIX4N_GOck:-LDIBtK3BUz5Y3z79SCQ:b32v57) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/search_tests.py#L65)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

Btw we should logout the clients once their utility is over. Else they will keep hogging http conncetion.
</blockquote></details>

+1

---


*Comments from [Reviewable](https://reviewable.io/reviews/vmware/pyvcloud/245#-:-LDIC4FJDIfERRPc7LiW:b95ijzd)*
<!-- Sent from Reviewable.io -->

hodgesrm commented:
Review status: 1 of 3 files reviewed at latest revision, 9 unresolved discussions.

---

*[system_tests/README.md, line 40 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDDSJtb5RrmV0IOI6oc:-LDHzuJ32KLiw8QbV23y:b-evjlc0) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/README.md#L40)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Can we mention about the naming convention as well for public test methods and private helper methods?
</blockquote></details>

Addressed. 

---

*[system_tests/README.md, line 51 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDDZsO4EwlVDE8NL7dB:-LDHzwSp89ur3HfBZaBW:b7bbvvc) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/README.md#L51)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

I guess he meant," if you _use_ unittest setup and teardown methods..."
</blockquote></details>

Clarified, thanks!

---

*[system_tests/README.md, line 67 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDDSUl3EreYW_iAhovT:-LDI4MXv8wVVFWvO0MCb:bkfhf25) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/README.md#L67)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

Yes sample base_config.yaml has been checked in git repo.
</blockquote></details>

Clarified in the description so this is mentioned earlier. 

---

*[system_tests/search_tests.py, line 23 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDEaLfS3E4sSLtO_GeR:-LDI7TMh14clPBY368m8:b5d8s8l) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/search_tests.py#L23)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

I think we should get rid of RESOURCE_TYPES and not use it. 
It's not being used anywhere else in the codebase. While introducing ResourceType enum, Curt kept RESOURCE_TYPE list just to be on the safe side.
</blockquote></details>

Removed from test. 

---

*[system_tests/search_tests.py, line 31 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDEaurf9QPcTkIlhXCR:-LDI5lxl80HKHFdmZeLl:b-j2llhx) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/search_tests.py#L31)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

This will run before every test. Is that the intent here?
</blockquote></details>

Yes.  It's cheap.  I would not do this for anything that sets vCD state.

---

*[system_tests/search_tests.py, line 48 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDEbS0nCTblirsv3wfI:-LDI7u4n7rVrdFwFdwlC:b-d98om6) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/search_tests.py#L48)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

Why not just use
name0 = q1_records[0] .name?
</blockquote></details>

Er, because I like more complicated code??? (Fixed, thanks!!)

---

*[system_tests/search_tests.py, line 65 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDIAJnU1NJIX4N_GOck:-LDIRNbmCmKoatWpTh4W:bkdtzhj) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/search_tests.py#L65)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Fixed, thanks.  I didn't logout the sysadmin connection for now but others are logged out using the Python unittest tearDown() method.  We should do this as a separate cleanup of tests which will need to happen anyway to get more of them officially running.

---

*[system_tests/search_tests.py, line 72 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDEcH1-5p9XEvQpFb7s:-LDI8O9k6hk0ChW8nvbk:bt2vr1b) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/search_tests.py#L72)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

.._find_existing_entities_..  ?
</blockquote></details>

Fixed.

---

*[system_tests/search_tests.py, line 110 at r2](https://reviewable.io/reviews/vmware/pyvcloud/245#-LDEdQ4q45OK-TBpJXMi:-LDI8TIr-4waTwGKPMvw:bea9umf) ([raw file](https://github.com/vmware/pyvcloud/blob/d652d674201eeb263182d817f2fd16aa509384d3/system_tests/search_tests.py#L110)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

Have you run this test locally?
We don't have resources like TO_CLOUD_TUNNEL in the testbed.
</blockquote></details>

Yes.  I have clarified in the code that the testing goal is to ensure the search does not fail but just returns something list-like (even if empty). 

---


*Comments from [Reviewable](https://reviewable.io/reviews/vmware/pyvcloud/245)*
<!-- Sent from Reviewable.io -->
